src/Mono.Android/Android.Runtime/ConstructorBuilder.cs

```diff
@@ -12,15 +12,14 @@ internal class ConstructorBuilder {
 		static FieldInfo handlefld = typeof (Java.Lang.Object).GetField ("handle", BindingFlags.NonPublic | BindingFlags.Instance);
 		static FieldInfo Throwable_handle = typeof (Java.Lang.Throwable).GetField ("handle", BindingFlags.NonPublic | BindingFlags.Instance);
 
-		static int dynamicMethodNameCounter;
 
 		internal static Action <IntPtr, object []> CreateDelegate (Type type, ConstructorInfo cinfo, Type [] parameter_types) {
 			var handle = handlefld;
 			if (typeof (Java.Lang.Throwable).IsAssignableFrom (type)) {
 				handle = Throwable_handle;
 			}
 
-			DynamicMethod method = new DynamicMethod ($"{Interlocked.Increment (ref dynamicMethodNameCounter)}c", typeof (void), new Type [] {typeof (IntPtr), typeof (object []) }, typeof (object), true);
+			DynamicMethod method = new DynamicMethod (DynamicMethodNameCounter.UniqueName, typeof (void), new Type [] {typeof (IntPtr), typeof (object []) }, typeof (DynamicMethodNameCounter), true);
```

Comment:
Do we really want constructors and methods to share the same counter?

I'm not even sure what range is "normal" for this counter, other than that the counter will increment for every constructor invocation and method registered.

Also note that -- at present -- the `ConstructorBuilder` count increments *every time `TypeManager.n_Activate()` is invoked*.  It's not incremented once per method signature; it's incremented once per *invocation*.  This is *entirely unbound*, and it's entirely imaginable to have Java code construct ~4 billion objects over the course of a (very) long-lived process, which would cause an overflow to occur.

Reply:
...that said, we still have that possible overflow problem *now*, because of PR #1872...and I didn't think of this scenario before.   Doh! :-(

Reply:
Is that really true? My understanding is that it is called once per type at most (many types don't have `<init>` constructors) and not per instance.

Reply:
> Is that really true?

Let's start with [Java Callable Wrappers](https://docs.microsoft.com/en-us/xamarin/android/platform/java-integration/android-callable-wrappers):

```java
    public HelloAndroid ()
    {
        super ();
        if (getClass () == HelloAndroid.class)
            mono.android.TypeManager.Activate (
                "Mono.Samples.HelloWorld.HelloAndroid, HelloWorld, Version=1.0.0.0, 
                Culture=neutral, PublicKeyToken=null", "", this, new java.lang.Object[] {  });
    }
```

`TypeManager.Activate()` is [`TypeManager.n_Activate()`](https://github.com/xamarin/xamarin-android/blob/77d603a2bd0778290d47b21b88fba5657df7e680/src/Mono.Android/Java.Interop/TypeManager.cs#L131-L177), and is invoked *separately* for *every* constructor invocation:

```java
// ~~~ From Java ~~~

new HelloAndroid();
// One TypeManager.n_Activate() invocation

new HelloAndroid();
// Two TypeManager.n_Activate() invocations

new HelloAndroid();
// Three TypeManager.n_Activate() invocations

...
```

Which brings us to the [`ConstructorBuilder.CreateDelegate()`](https://github.com/xamarin/xamarin-android/blob/77d603a2bd0778290d47b21b88fba5657df7e680/src/Mono.Android/Java.Interop/TypeManager.cs#L167) invocation, for which *nothing is cached*.  (We should *absolutely* add some caching; it would have noticeable improvements in instance construction!  However, we don't *currently* have any such caching.)

Which is why I said:

> the ConstructorBuilder count increments *every time `TypeManager.n_Activate()` is invoked*.

I don't see any other interpretation.

Reply:
Interesting, I will need to take look at it. Hopefully we can address that in https://github.com/xamarin/xamarin-android/issues/2022.

Meanwhile I have updated the patch so, that we get back to using *Guid* once we create `int.MaxValue` methods. I am curious whether it would even be possible to associate that many methods with a type :-) Didn't have time to try it yet.

Reply:
>  I am curious whether it would even be possible to associate that many methods with a type :-)

It isn't possible to associate int.MaxValue methods with a type; the Java `.class` file format only permits *`ushort.MaxValue`* values within the "constant pool" of a `.class` file, and each method declared within a Java type must have one or more entries in the constant pool.

However, `dynamicMethodNameCounter` isn't "scoped" to a single type; it's used across *all* types.

If it were limited to *just* methods and not constructors, this means the developer would need to register `int.MaxValue/ushort.MaxValue`=32768 types in order to overflow the counter.

This is unlikely. :-)

The problem, again, is constructors: constructors are not per-type, they're per-*invocation*.  If Java had an equivalent to [`ConstructorInfo.Invoke(object, object[])`](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.methodbase.invoke?view=netframework-4.7.2#System_Reflection_MethodBase_Invoke_System_Object_System_Object___) -- which perhaps might not exist in Java, but let's go with it -- you would only need to initialize *one* type, and one instance, and just call that one method int.MaxValue times.

That's excessive, but seems considerably more likely than having 32768 types in a `.apk`.

Reply:
I mean the type on the managed side here, where the methods/constructors are associated to `DynamicMethodCounter` class (or System.Object as before).